2025 08 18 fuse box puzzle

JosephStar318 opened the pull request:
Added fuse box puzzle and implemented generalized puzzle system for future uses.

Commit:
- puzzle controller, base puzzle and quest puzzle notifier added

- fixed a bug where interaction raycaster would not hide the hint when getting far away

Commit:
- puzzle save and load system integrated

- once a puzzle is solved interaction is blocked. this can be overriden if necesserray in the derived puzzle
- quest puzzle notifier notifies the key "puzzle_{puzzleID}"

Commit:
- phantom camera addon imported

Commit:
- fusebox puzzle prototype ready

- fixed management of handheld device when entering a puzzle

Commit:
- puzzle placement changed

- name and description added

JosephStar318 commented:
BasePuzzle has 3 major components. First is PuzzleData which is like a metadata for the puzzle that contains id name and description. 
This puzzleData also needs to be in the ResourceDatabase for the save system to work properly.

Labels is the information that will be visible once the camera finishes tweening and invisible once the player backs from the puzzle. You can add any information on how to play the puzzle or KeybindLabel3D to show which button does what.

Phantom Camera is the puzzles camera. When you interact with the puzzle main camera will automatically switch to this camera with the tween parameters provided inside the phantom camera

Also since this inherits from Interactable, it will require a Hint node. You can copy paste from any other interactable in the game.

InteractionRaycaster will automatically detect the Interactable and draw the hint so there is no need to handle input manually.

Also the BasePuzzle will provide predetermined inputs like wasd, up down left right arrows for movement, enter for submit,  R for reset, C for back.

Once the puzzle is solved we will trigger SetSolved(true) in the script. it will handle the events and auto backs from the puzzle.
You can check out FuseBoxPuzzle.cs for general idea of how to implement your puzzles.

<img width="2048" height="370" alt="image" src="https://github.com/user-attachments/assets/a9ae0891-15b0-40d4-842f-74942bfd8b34" />

JosephStar318 commented:
I have also added support for quests

For this you need to add a QuestPuzzleNotifier script to any node and bind the puzzle to it. Whenever a puzzle is solved it will set the quest variable of “puzzle_id” to true. Id is inside PuzzleData.

Commit:
- puzzle controller input vector bug fix

JosephStar318 committed:
Merge branch 'staging' into 2025-08-18-FuseBoxPuzzle

TheKingChu committed:
Lift puzzle works with puzzle system

TheKingChu reviewed (approved):
Think it should all be working together now!

JosephStar318 commented on scripts/Puzzles/LiftTestPuzzle.cs:
```diff
@@ -1,80 +1,49 @@
-        if(isSolved) return;
+        if(Data.IsSolved) return;
 
         if(Godot.Input.IsActionJustPressed("interact") && IsPlayerInRange())
```
You don't need to read interact input. Puzzles will receive Interact() callback automatically. If you want to set a requirement like player should be in range, you can override CanInteract(). And if you add a hint it will display the hint when in range as well

JosephStar318 replied:
This is only valid if you are looking at the puzzle though. If that's the case and you don't want to be looking at the puzzle for interaction to work, current state is good

JosephStar318 replied:
i think i was misunderstood where i said you don't need to read interact input. I was meaning that you don't need to actually call Interact(). If you look at the InteractionRaycaster.cs it checks a proximity and do a CanInteract(). If everything is good it handles the keyboard event and calls the Interact(). So just by overriding the interact you already read input. 

But there is a catch which i don't know. If you made it so that where you don't want to be looking at the object itself to interact with, current system can't understand where to interact so its okay for you to have an interact input check if that's the case.

JosephStar318 replied:
I hope i was more clear this time 😅

JosephStar318 reviewed (commented):
some comments added

TheKingChu committed:
Fixed up from comment

Removed listening for interaction and just look for player, also added comment for ovrriding interact

Commit:
- interaction raycaster can detect the parent node
